File: Data Structures & Algorithms/binary-tree-diameter/submission-6.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func diameterOfBinaryTree(root *TreeNode) int {
	maxDiameter := 0
	var dfs func(root *TreeNode) int
	dfs = func(root *TreeNode) int {
		// base case 
		if root == nil {
			return 0
		}
		// якщо діаметер це сума макс ребр, то це виходить треба висота наступної ноди
		// leap of faith -> мій dfs рахує висоту поточного вузла і перевіряє чи більший діаметер за попереднє значення
		// висота вузла - це к-сть вузлів від найнижчого листка до поточного вузла
		leftHeight := dfs(root.Left)
		rightHeight := dfs(root.Right)
		
		diameter := leftHeight + rightHeight
		maxDiameter = max(maxDiameter, diameter)

		return 1 + max(leftHeight, rightHeight)
	}

	dfs(root)
	return maxDiameter
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSameTree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        
        if not root or not subRoot:
            return False
        
        isValid = root.val == subRoot.val
        isLeftValid = self.isSameTree(root.left, subRoot.left)
        isRightValid = self.isSameTree(root.right, subRoot.right)

        return isValid and isLeftValid and isRightValid

                
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode], subroot: Optional[TreeNode]) -> bool:
            # base case - пустий корінь
            if not root:
                return False
            
            isSameTree = self.isSameTree(root, subroot)
            isLeftSame = dfs(root.left, subroot)
            isRightSame = dfs(root.right, subroot)

            return isSameTree or isLeftSame or isRightSame

        return dfs(root, subRoot)
